pkg/webhook/resources/storageclass/validator.go

```diff
@@ -0,0 +1,77 @@
+package storageclass
+
+import (
+	ctlstoragev1 "github.com/rancher/wrangler/pkg/generated/controllers/storage/v1"
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/harvester/harvester/pkg/util"
+	werror "github.com/harvester/harvester/pkg/webhook/error"
+	"github.com/harvester/harvester/pkg/webhook/types"
+)
+
+func NewValidator(storageClassCache ctlstoragev1.StorageClassCache) types.Validator {
+	return &storageClassValidator{
+		storageClassCache: storageClassCache,
+	}
+}
+
+type storageClassValidator struct {
+	types.DefaultValidator
+	storageClassCache ctlstoragev1.StorageClassCache
+}
+
+func (v *storageClassValidator) Resource() types.Resource {
+	return types.Resource{
+		Names:      []string{"storageclasses"},
+		Scope:      admissionregv1.ClusterScope,
+		APIGroup:   storagev1.SchemeGroupVersion.Group,
+		APIVersion: storagev1.SchemeGroupVersion.Version,
+		ObjectType: &storagev1.StorageClass{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Create,
+			admissionregv1.Update,
+		},
+	}
+}
+
+func (v *storageClassValidator) Create(request *types.Request, newObj runtime.Object) error {
+	return v.validateSetUniqueDefault(newObj)
+}
+
+func (v *storageClassValidator) Update(request *types.Request, oldObj runtime.Object, newObj runtime.Object) error {
+	return v.validateSetUniqueDefault(newObj)
+}
+
+func (v *storageClassValidator) validateSetUniqueDefault(newObj runtime.Object) error {
+	newSC := newObj.(*storagev1.StorageClass)
+
+	scList, err := v.storageClassCache.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	return validateSetUniqueDefault(newSC, scList)
+}
+
+func validateSetUniqueDefault(newSC *storagev1.StorageClass, scList []*storagev1.StorageClass) error {
+	// if new storage class haven't storageclass.kubernetes.io/is-default-class annotation or the value is false.
+	if v, ok := newSC.Annotations[util.AnnotationIsDefaultStorageClassName]; !ok || v == "false" {
+		return nil
+	}
+
```

Comment:
prefer to move v.storageClassCache.List logic to here

and merge two validateSetUniqueDefault function